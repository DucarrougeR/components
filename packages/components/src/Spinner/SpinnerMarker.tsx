import { quarterFade } from 'looker-design-tokens'
import styled, { css } from 'styled-components'
import { color, ColorProps } from 'styled-system'

interface SpinnerMarkerProps extends ColorProps {
  markers: number
  markerIndex: number
  markerRadius?: number
  speed: number
}

const markerStyle = (props: SpinnerMarkerProps) => {
  const { markerIndex, markerRadius, markers, speed } = props
  const delay = (markerIndex * speed) / markers
  const rotateAngle = (360 / markers) * markerIndex

  return css`
    animation: ${quarterFade} ${speed}ms linear ${delay}ms infinite;
    border-radius: ${markerRadius && `${markerRadius}px`};
    transform: rotate(${rotateAngle}deg) translate(0, -160%);
  `
}

export const SpinnerMarker = styled.div<SpinnerMarkerProps>`
  ${color}
  ${markerStyle}
  height: 20%;
  left: 48%;
  opacity: 0.25;
  position: absolute;
  top: 40%;
  width: 6%;
`
