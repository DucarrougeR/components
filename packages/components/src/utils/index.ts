export * from './GlobalStyle'
export * from './useToggle'
