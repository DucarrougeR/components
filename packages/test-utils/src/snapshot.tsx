import 'jest-styled-components'
import React from 'react'
import { createWithTheme, shallowWithTheme } from './create_with_theme'

export const assertSnapshot = <T extends {}>(
  element: React.ReactElement<T>
) => {
  expect(createWithTheme(element).toJSON()).toMatchSnapshot()
}

export const assertSnapshotShallow = <T extends {}>(
  element: React.ReactElement<T>
) => {
  expect(shallowWithTheme(element)).toMatchSnapshot()
}
