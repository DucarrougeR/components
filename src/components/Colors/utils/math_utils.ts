/**
 * Utility to convert a numerical value to a percent given the value and a max value.
 */
export const toPerc = (value: number, maxValue: number) =>
  Math.round((value / maxValue) * 100)
