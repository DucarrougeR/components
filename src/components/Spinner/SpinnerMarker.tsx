import * as React from 'react'
import { css, styled } from '../../style'
import { Box, BoxPropsWithout } from '../Box'

export interface SpinnerMarkerProps
  extends BoxPropsWithout<HTMLDivElement, 'color'> {
  color?: string
  radius?: number
  rotateAngle: number
  circleSize: number
  delay: number
  speed: number
}

const InternalMarker: React.FC<SpinnerMarkerProps> = ({
  color,
  radius,
  rotateAngle,
  circleSize,
  delay,
  speed,
  ...props
}) => <Box bg={color} borderRadius={radius} {...props} />

const markerTransform = (props: SpinnerMarkerProps) => {
  return css`
    transform: rotate(${props.rotateAngle}deg) translate(0, -160%);
    animation-delay: ${props.delay}ms;
  `
}

const SpinnerMarkerFactory = React.forwardRef(
  (props: SpinnerMarkerProps, ref) => (
    <InternalMarker innerRef={ref} {...props} />
  )
)

export const SpinnerMarker = styled<SpinnerMarkerProps>(SpinnerMarkerFactory)`
  animation-duration: ${props => props.speed + 'ms'};
  width: 6%;
  height: 20%;
  position: absolute;
  left: 48%;
  top: 40%;
  opacity: 0.25;
  ${markerTransform};
`
