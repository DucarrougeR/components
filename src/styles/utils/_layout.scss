// Returns a value from the $spacing list
@function spacing($key) {
  @return map-lookup($spacing, $key, '$spacing');
}
