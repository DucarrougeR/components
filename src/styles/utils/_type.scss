// Returns a value from the $test-size list
@function font-size($key) {
  @return map-lookup($text-size, $key, '$text-size');
 }

 // Returns a value from the $line-height list
 @function line-height($key) {
   @return map-lookup($line-height, $key, '$line-height');
 }

 @mixin fontSize($size) {
  font-size: $size;
  font-size: pxToRem($size);
}
