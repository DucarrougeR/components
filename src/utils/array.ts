/**
 * Returns an array of length `size` filled with zeros.
 * Useful for looping constructs where a length is known, but no array is provided.
 */

export const sizedArray = (size: number) => Array(size).fill(0)
