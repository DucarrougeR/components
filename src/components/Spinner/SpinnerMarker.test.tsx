import 'jest-styled-components'
import * as React from 'react'
import { assertSnapshot } from '../../../test/utils/snapshot'
import { SpinnerMarker } from './SpinnerMarker'

describe('A SpinnerMarker', () => {
  test('A default SpinnerMarker', () => {
    assertSnapshot(
      <SpinnerMarker
        rotateAngle={20}
        circleSize={200}
        speed={1000}
        delay={200}
      />
    )
  })

  test('A SpinnerMarker radius can be changed', () => {
    assertSnapshot(
      <SpinnerMarker
        rotateAngle={20}
        circleSize={200}
        speed={1000}
        delay={200}
        radius={200}
      />
    )
  })

  test('A SpinnerMarker color can be changed', () => {
    assertSnapshot(
      <SpinnerMarker
        rotateAngle={20}
        circleSize={200}
        speed={1000}
        delay={200}
        color={'blue'}
      />
    )
  })
})
