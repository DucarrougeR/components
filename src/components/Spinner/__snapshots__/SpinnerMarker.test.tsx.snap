// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A SpinnerMarker A SpinnerMarker color can be changed 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  background-color: blue;
}

.c0 {
  -webkit-animation-duration: 1000ms;
  animation-duration: 1000ms;
  width: 6%;
  height: 20%;
  position: absolute;
  left: 48%;
  top: 40%;
  opacity: 0.25;
  -webkit-transform: rotate(20deg) translate(0,-160%);
  -ms-transform: rotate(20deg) translate(0,-160%);
  transform: rotate(20deg) translate(0,-160%);
  -webkit-animation-delay: 200ms;
  animation-delay: 200ms;
}

<div
  className="c0 c1"
/>
`;

exports[`A SpinnerMarker A SpinnerMarker radius can be changed 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 200px;
}

.c0 {
  -webkit-animation-duration: 1000ms;
  animation-duration: 1000ms;
  width: 6%;
  height: 20%;
  position: absolute;
  left: 48%;
  top: 40%;
  opacity: 0.25;
  -webkit-transform: rotate(20deg) translate(0,-160%);
  -ms-transform: rotate(20deg) translate(0,-160%);
  transform: rotate(20deg) translate(0,-160%);
  -webkit-animation-delay: 200ms;
  animation-delay: 200ms;
}

<div
  className="c0 c1"
/>
`;

exports[`A SpinnerMarker A default SpinnerMarker 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c0 {
  -webkit-animation-duration: 1000ms;
  animation-duration: 1000ms;
  width: 6%;
  height: 20%;
  position: absolute;
  left: 48%;
  top: 40%;
  opacity: 0.25;
  -webkit-transform: rotate(20deg) translate(0,-160%);
  -ms-transform: rotate(20deg) translate(0,-160%);
  transform: rotate(20deg) translate(0,-160%);
  -webkit-animation-delay: 200ms;
  animation-delay: 200ms;
}

<div
  className="c0 c1"
/>
`;
